pythran/analyses/aliases.py

```diff
@@ -363,38 +569,118 @@ def merge_return_aliases(args):
+
+        Everyone points to the formal parameter 'a' \o/
+        '''
         md.visit(self, node)
         value_aliases = self.visit(node.value)
         for t in node.targets:
             if isinstance(t, ast.Name):
-                self.aliases[t.id] = value_aliases or {t}
+                self.aliases[t.id] = set(value_aliases) or {t}
                 for alias in list(value_aliases):
                     if isinstance(alias, ast.Name):
-                        self.aliases[alias.id].add(t)
-                self.add(t, self.aliases[t.id].copy())
+                        a_id = alias.id
+                        self.aliases[a_id] = self.aliases[a_id].union((t,))
+                self.add(t, self.aliases[t.id])
             else:
                 self.visit(t)
 
     def visit_For(self, node):
-        # FIXME: node.target.id could alias to the content of node.iter
-        self.aliases[node.target.id] = {node.target}
+        '''
+        For loop creates aliasing between the target
+        and the content of the iterator
+
+        >>> from pythran import passmanager
+        >>> pm = passmanager.PassManager('demo')
+        >>> module = ast.parse("""
+        ... def foo(a):
+        ...     for i in a:
+        ...         {i}""")
+        >>> result = pm.gather(Aliases, module)
+        >>> Aliases.dump(result, filter=ast.Set)
+        {i} => ['|i|']
+
+        Not very useful, unless we know something about the iterated container
+
+        >>> module = ast.parse("""
+        ... def foo(a, b):
+        ...     for i in [a, b]:
+        ...         {i}""")
+        >>> result = pm.gather(Aliases, module)
+        >>> Aliases.dump(result, filter=ast.Set)
+        {i} => ['|a|', '|b|']
+        '''
+
+        iter_aliases = self.visit(node.iter)
+        if all(isinstance(x, ContainerOf) for x in iter_aliases):
+            target_aliases = set()
+            for iter_alias in iter_aliases:
+                target_aliases.add(iter_alias.containee)
+        else:
+            target_aliases = {node.target}
+
+        self.add(node.target, target_aliases)
+        self.aliases[node.target.id] = self.result[node.target]
         # Error may come from false branch evaluation so we have to try again
         try:
             self.generic_visit(node)
         except PythranSyntaxError:
             self.generic_visit(node)
 
     def visit_While(self, node):
+        '''
+
+        While statement evaluation is somehow equivalent to the evaluation of a
+        sequence, except the fact that in some subtle cases, the first rounds
+        of analyse fails because we do not follow the regular execution order
+
+        >>> from pythran import passmanager
+        >>> pm = passmanager.PassManager('demo')
+        >>> fun = """
+        ... def foo(a):
+        ...     while(a):
+        ...         if a==1: print b
+        ...         else: b=a"""
+        >>> module = ast.parse(fun)
+        >>> result = pm.gather(Aliases, module)
```

Comment:
No, I don't know what to check here... maybe return b and check the return_alias?